Restore indexes from backup with the original partitioning

jepett0 opened the pull request:
https://github.com/ydb-platform/ydb/issues/7608

This PR enables restoring indices from backups with the original partition split boundaries.

This feature is mostly needed to speed-up the restoration from backups. According to my tests (tpch `lineitem` table with 156x scale on an 8 node cluster with the `cpu80_soc2_mem512G_net25G_4ssd` preset), it cuts the duration of the `BuildIndexes` stage of the `import/s3` operation from `829` seconds to `568` seconds when restoring a 150GiB table  with a single 100GiB index. This is a 31% reduction in `BuildIndexes` time! 🎉 Total restoration from backup time went from `1259` seconds to `995` seconds which is a 21% reduction.

C++ SDK is also [changed a little](https://github.com/ydb-platform/ydb/pull/7589/files#diff-4ac938110466e9b5daab6438f5594540e857bea4b1556721d26491fa3cbae2a5) to enable users to create a table with an index that has specific partitioning settings and uniform partition count or explicit split boundaries. Enabling TTableBuilder to add an index based on its description makes it as capable as session.AlterTable [already was](https://github.com/ydb-platform/ydb/blob/8a1999fc4070371153b1a5b16fec99e0d212d6de/ydb/public/sdk/cpp/client/ydb_table/table.h#L1536). You can see how it can be helpful in the added [test](https://github.com/ydb-platform/ydb/blob/8a1999fc4070371153b1a5b16fec99e0d212d6de/ydb/services/ydb/backup_ut/ydb_backup_ut.cpp#L282).

jepett0 commented on ydb/core/tx/schemeshard/schemeshard_path_describer.cpp:
```diff
@@ -499,7 +499,7 @@ void TPathDescriber::DescribeTable(const TActorContext& ctx, TPathId pathId, TPa
         switch (childPath->PathType) {
         case NKikimrSchemeOp::EPathTypeTableIndex:
-            Self->DescribeTableIndex(childPathId, childName, returnConfig, false, *entry->AddTableIndexes());
+            Self->DescribeTableIndex(childPathId, childName, returnConfig, returnBoundaries, *entry->AddTableIndexes());
```
This is the most problematic change.

We inherit `returnBoundaries` from the parent table's describe options and return indexImplTable boundaries even if the user is interested only in the main table boundaries. It might induce unnecessary message passing between actors in YDB especially since indexImplTables might have a very large number of partitions.

One way to solve this issue is to add a separate describe table [parameter](https://github.com/ydb-platform/ydb/blob/6b9001c607447a9e94cae8f4c3fbab768463019a/ydb/core/protos/flat_scheme_op.proto#L1834) to control the indexImplTable boundaries retrieval.

jepett0 replied:
Fixed in the latest [revision](https://github.com/ydb-platform/ydb/pull/7589/commits/33900bb676aeff4ce3702507a12590c901554f24)

MBkkt commented on ydb/core/tx/schemeshard/schemeshard_path_describer.cpp:
```diff
@@ -72,6 +72,79 @@ static void FillTableStats(NKikimrSchemeOp::TPathDescription& pathDescription, c
+                familyNamesBuilt = true;
+            }
+
+            auto it = familyNames.find(cinfo.Family);
+            if (it != familyNames.end() && !it->second.empty()) {
+                colDescr->SetFamilyName(it->second);
```
NIT:
FindPtr probably more convinient here

jepett0 replied:
I have copy-pasted this code. Don't want to dig into it

jepett0 committed:
refactoring to reduce (future) code duplication

jepett0 committed:
bugfix

This method was never called, so this bug was hidden.

jepett0 committed:
enable indexImplTable split boundaries export

+ Add a dedicated option to control inclusion of the indexImplTable boundaries in the main table description. This should help minimizing the network traffic and IO CPU pool usage.

jepett0 committed:
supporting changes in SDK

These are needed for a good looking test. However, these functions might be helpful for users also. It is currently the only way (except direct GRPCs) to create a table with an index that has predefined split boundaries.

jepett0 committed:
refactoring in tests

jepett0 committed:
add new tests

CyberROFL commented on ydb/public/api/protos/ydb_table.proto:
```diff
@@ -773,12 +773,16 @@ message DescribeTableRequest {
+    // Note: The following flag applies only if the described table has indices.
+    // Includes partition boundaries (the key values where the partitions are split)
+    // for tables that implement indices of the described table.
+    bool include_index_table_shard_key_bounds = 8;
```
Имеет ли смысл комбинация `include_shard_key_bounds: false` и `include_index_table_shard_key_bounds: true`?

spuchin replied:
Может в случае include_partition_stats=true просто возвращать и для таблицы и для индексов?

jepett0 replied:
> Может в случае include_partition_stats=true просто возвращать и для таблицы и для индексов?

Да, такая идея была, но мы с @ijon посчитали, что правильнее всё-таки сделать выделенный флажок. Не всякий describe с include_partition_stats=true нуждается ещё и в границах партиций индексных таблиц. А overhead от пересылки границ партиций indexImplTables может превосходить нагрузку от пересылки границ партиций родительской таблицы, потому что партиций у индексных таблиц суммарно может быть сильно больше.

jepett0 replied:
> Имеет ли смысл комбинация `include_shard_key_bounds: false` и `include_index_table_shard_key_bounds: true`?

Скорее, нет. Но вот
`include_shard_key_bounds: true` && `include_index_table_shard_key_bounds: false` выглядит вполне осмысленно и пригодно.

Не знаю, как тут обойтись без отдельного флажка

jepett0 replied:
Есть идея, как избавиться от отсутствия смысла в сочетании:
- `include_shard_key_bounds: false`
- `include_index_table_shard_key_bounds: true`

Мы можем переделать флажок `include_index_table_shard_key_bounds` во что-то в духе:
`include_extended_index_info`
и применять его в сочетании с существующими:
- `include_shard_key_bounds`
- `include_partition_stats`

Захотел пользователь увидеть границы партиций индексной таблицы, глядя на describe родительской таблицы? Ok, пусть включает и `include_shard_key_bounds` и `include_extended_index_info`.

Захотел увидеть информацию о партиционировании индексной таблицы, глядя на describe родительской таблицы? Ok, пусть включит и `include_partition_stats` и `include_extended_index_info`.

jepett0 replied:
Удалил новую опцию `include_index_table_shard_key_bounds` в последней ревизии.

`ydb export` (это backup в S3) будет работать и без неё. Она нужна только для более старого локального backup'a с помощью `ydb tools dump`. Её мы доработаем позже, в рамка отдельного [тикета](https://github.com/ydb-platform/ydb/issues/8054).

jepett0 committed:
review fix #1

- Change signature of AddSecondaryIndex function to receive TIndexDescription. It should be much more future proof. Besides TAlterTableSettings already supports adding indexes using TIndexDescription. This change will unify TTableBuilder and TAlterTableSettings in this regard.

jepett0 committed:
review fix #2

- remove the new option to control index impl tables partition split boundaries inclusion in the describe result of a table from public API. We will work on `ydb tools dump` later to enable it to save index tables split boundaries.
